File: mobile/dominio/prestadores/validador-prestador.ts
/**
 * Validador y factory del módulo PRESTADORES — reglas multi-tenant.
 *
 * El módulo expone:
 *   - `cedulaRepresentanteLegalValida` — regla pura `boolean` (Fase 2.1).
 *   - `crearPrestador` — factory que valida TODAS las reglas de creación
 *     y devuelve un `PrestadorBorrador` listo para persistir.
 *
 * Las funciones puras `boolean` se usan desde la UI para feedback inline
 * sin acoplar a strings de dominio. La factory, en cambio, encapsula el
 * "shape correcto" de un prestador nuevo y centraliza el catálogo de
 * errores `MENSAJES_ERROR_PRESTADOR` (mirror del patrón de
 * `dominio/operarios/operarios.ts` → `crearOperario`).
 */

import type { CrearPrestadorInput, PrestadorBorrador } from './types';
import { MENSAJES_ERROR_PRESTADOR } from './types';

const REGEX_CEDULA_COLOMBIA = /^\d{6,12}$/;
const REGEX_CODIGO = /^\d{1,50}$/;
const LONGITUD_NOMBRE_MAXIMA = 200;
const LONGITUD_NIT_MAXIMA = 20;
const LONGITUD_MUNICIPIO_MAXIMA = 100;
const LONGITUD_DEPARTAMENTO_MAXIMA = 100;

/**
 * Valida la cédula del representante legal del prestador.
 *
 * Acepta exclusivamente cadenas de dígitos entre 6 y 12 caracteres. No
 * admite separadores (puntos, guiones, espacios) ni letras; la regla
 * busca el formato canónico colombiano que se almacena en la columna
 * `representante_legal_cedula` agregada por migration 016.
 *
 * @param cedula cadena cruda del input del usuario.
 * @returns `true` si cumple el formato, `false` en cualquier otro caso
 *          (incluye string vacío, `undefined` no convertible, longitudes
 *          fuera de [6, 12], separadores o caracteres no numéricos).
 */
export function cedulaRepresentanteLegalValida(cedula: string): boolean {
  return REGEX_CEDULA_COLOMBIA.test(cedula);
}

/**
 * Valida todas las reglas de creación de un prestador.
 *
 * Lanza `Error` con la clave de `MENSAJES_ERROR_PRESTADOR` que corresponda
 * a la primera violación encontrada. El orden de las comprobaciones es
 * estable para que los tests (y la UI) puedan anticipar el mensaje.
 *
 * Reglas:
 *   - codigo: 1-50 dígitos (`/^\d{1,50}$/`).
 *   - nombre: 1-200 caracteres.
 *   - nit: 1-20 caracteres (chequeo de formato queda fuera del MVP rural).
 *   - representante_legal: no vacío.
 *   - representante_legal_cedula: 6-12 dígitos (reusa el validador
 *     `cedulaRepresentanteLegalValida` de Fase 2.1).
 *   - municipio: 1-100 caracteres.
 *   - departamento: 1-100 caracteres.
 *   - segmento: estrictamente 1 o 2.
 *   - num_suscriptores_urbanos / rurales: enteros ≥ 0.
 */
function validarEntrada(input: CrearPrestadorInput): void {
  if (!REGEX_CODIGO.test(input.codigo)) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.CODIGO_VACIO);
  }
  if (input.nombre.length === 0) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.NOMBRE_VACIO);
  }
  if (input.nombre.length > LONGITUD_NOMBRE_MAXIMA) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.NOMBRE_LARGO);
  }
  if (input.nit.length === 0) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.NIT_VACIO);
  }
  if (input.nit.length > LONGITUD_NIT_MAXIMA) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.NIT_LARGO);
  }
  if (input.representante_legal.length === 0) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.REPRESENTANTE_LEGAL_VACIO);
  }
  if (!cedulaRepresentanteLegalValida(input.representante_legal_cedula)) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.CEDULA_REP_LEGAL_INVALIDA);
  }
  if (input.municipio.length === 0) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.MUNICIPIO_VACIO);
  }
  if (input.municipio.length > LONGITUD_MUNICIPIO_MAXIMA) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.MUNICIPIO_LARGO);
  }
  if (input.departamento.length === 0) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.DEPARTAMENTO_VACIO);
  }
  if (input.departamento.length > LONGITUD_DEPARTAMENTO_MAXIMA) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.DEPARTAMENTO_LARGO);
  }
  if (input.segmento !== 1 && input.segmento !== 2) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.SEGMENTO_INVALIDO);
  }
  if (input.num_suscriptores_urbanos < 0) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.NUM_URBANOS_NEGATIVO);
  }
  if (input.num_suscriptores_rurales < 0) {
    throw new Error(MENSAJES_ERROR_PRESTADOR.NUM_RURALES_NEGATIVO);
  }
}

/**
 * Crea un `PrestadorBorrador` a partir de un `CrearPrestadorInput`,
 * aplicando defaults (`estado = 'activo'`, `contacto = null`) si el
 * caller los omitió y validando el resto de las reglas de negocio.
 *
 * No aplica unicidad de `codigo` ni persistencia: esa es responsabilidad
 * del repositorio SQLite (que ya consulta la regla `existePorCodigo`
 * antes de delegar acá).
 *
 * @throws `Error` con `MENSAJES_ERROR_PRESTADOR.X` ante la primera
 *         violación de las reglas enumeradas en `validarEntrada`.
 */
export function crearPrestador(input: CrearPrestadorInput): PrestadorBorrador {
  validarEntrada(input);

  return {
    codigo: input.codigo,
    nombre: input.nombre,
    nit: input.nit,
    representante_legal: input.representante_legal,
    representante_legal_cedula: input.representante_legal_cedula,
    municipio: input.municipio,
    departamento: input.departamento,
    segmento: input.segmento,
    num_suscriptores_urbanos: input.num_suscriptores_urbanos,
    num_suscriptores_rurales: input.num_suscriptores_rurales,
    contacto: input.contacto ?? null,
    estado: input.estado ?? 'activo',
  };
}

File: backend/src/MediApp.Api/Dominio/Validacion/ValidadorPrestador.cs
using System.Text.RegularExpressions;

namespace MediApp.Api.Dominio.Validacion;

/// <summary>
/// Validador puro del módulo PRESTADORES — reglas multi-tenant.
///
/// Espejo simétrico del validador mobile
/// <c>mobile/dominio/prestadores/validador-prestador.ts</c>. Las funciones
/// retornan <c>bool</c> para mantener el validador testeable sin acoplar a
/// strings de UI ni a FluentValidation (la validación de payload en endpoints
/// sigue delegando a <c>Features/Prestadores/PrestadorValidator.cs</c>).
///
/// Aplicar ANTES de persistir:
///   - <c>EsCodigoValido</c> → 1-50 dígitos (<c>^\d{1,50}$</c>).
///   - <c>EsNombreValido</c> → 1-200 caracteres.
///   - <c>EsNitValido</c> → 1-20 caracteres.
///   - <c>EsRepresentanteLegalValido</c> → no vacío.
///   - <c>EsCedulaRepresentanteLegalValida</c> → 6-12 dígitos (<c>^\d{6,12}$</c>,
///     mismo regex que mobile <c>cedulaRepresentanteLegalValida</c>).
///   - <c>EsMunicipioValido</c> / <c>EsDepartamentoValido</c> → 1-100 caracteres.
///   - <c>EsSegmentoValido</c> → estrictamente 1 o 2 (Res CRA 825/2017 art. 6).
///   - <c>EsNumSuscriptoresUrbanosValido</c> / <c>EsNumSuscriptoresRuralesValido</c> → ≥ 0.
///
/// Decisión 2026-07-09: <c>EsCodigoValido</c> usa regex más estricto
/// (<c>^\d{1,50}$</c>) que el <c>PrestadorValidator</c> FluentValidation
/// preexistente (<c>^[A-Za-z0-9\-]{1,50}$</c>). Esto replica el contrato
/// del mobile para garantizar simetría cross-platform. Ambos coexisten.
/// </summary>
public static class ValidadorPrestador
{
    private const int LONGITUD_NOMBRE_MAXIMA = 200;
    private const int LONGITUD_NIT_MAXIMA = 20;
    private const int LONGITUD_MUNICIPIO_MAXIMA = 100;
    private const int LONGITUD_DEPARTAMENTO_MAXIMA = 100;

    private static readonly Regex CodigoRegex =
        new(@"^\d{1,50}$", RegexOptions.Compiled);

    private static readonly Regex CedulaRepresentanteLegalRegex =
        new(@"^\d{6,12}$", RegexOptions.Compiled);

    /// <summary>
    /// Acepta exclusivamente 1-50 dígitos numéricos. Sin letras, guiones,
    /// puntos ni espacios — mismo formato canónico que el regex mobile
    /// <c>REGEX_CODIGO = /^\d{1,50}$/</c>.
    /// </summary>
    public static bool EsCodigoValido(string? codigo)
    {
        return codigo is not null && CodigoRegex.IsMatch(codigo);
    }

    /// <summary>1-200 caracteres, no nulo, no whitespace-puro lo acepta (puede ser nombre válido).</summary>
    public static bool EsNombreValido(string? nombre)
    {
        if (string.IsNullOrEmpty(nombre)) return false;
        return nombre.Length <= LONGITUD_NOMBRE_MAXIMA;
    }

    /// <summary>1-20 caracteres, no nulo.</summary>
    public static bool EsNitValido(string? nit)
    {
        if (string.IsNullOrEmpty(nit)) return false;
        return nit.Length <= LONGITUD_NIT_MAXIMA;
    }

    /// <summary>No vacío. La longitud se delega a DataAnnotations de la entidad.</summary>
    public static bool EsRepresentanteLegalValido(string? valor)
    {
        return !string.IsNullOrEmpty(valor);
    }

    /// <summary>
    /// Cédula del representante legal (6-12 dígitos). Mismo regex que el
    /// validador mobile <c>cedulaRepresentanteLegalValida</c>.
    /// </summary>
    public static bool EsCedulaRepresentanteLegalValida(string? cedula)
    {
        return cedula is not null && CedulaRepresentanteLegalRegex.IsMatch(cedula);
    }

    /// <summary>1-100 caracteres, no nulo.</summary>
    public static bool EsMunicipioValido(string? municipio)
    {
        if (string.IsNullOrEmpty(municipio)) return false;
        return municipio.Length <= LONGITUD_MUNICIPIO_MAXIMA;
    }

    /// <summary>1-100 caracteres, no nulo.</summary>
    public static bool EsDepartamentoValido(string? departamento)
    {
        if (string.IsNullOrEmpty(departamento)) return false;
        return departamento.Length <= LONGITUD_DEPARTAMENTO_MAXIMA;
    }

    /// <summary>Estrictamente 1 o 2. Cualquier otro valor → false.</summary>
    public static bool EsSegmentoValido(short segmento)
    {
        return segmento == 1 || segmento == 2;
    }

    /// <summary>Entero ≥ 0.</summary>
    public static bool EsNumSuscriptoresUrbanosValido(int n)
    {
        return n >= 0;
    }

    /// <summary>Entero ≥ 0.</summary>
    public static bool EsNumSuscriptoresRuralesValido(int n)
    {
        return n >= 0;
    }
}
